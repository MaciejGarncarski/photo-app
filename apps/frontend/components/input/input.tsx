import clsx from 'clsx'
import { type ChangeEvent, Ref } from 'react'

import styles from './input.module.scss'

type Props = {
	labelText: string
	placeholder: string
	type?: 'text' | 'number' | 'tel' | 'email' | 'password'
	variant: 'primary' | 'secondary'
	error?: string
	optional?: boolean
	value?: string
	ref?: Ref<HTMLInputElement>
	onChange?: (changeEv: ChangeEvent<HTMLInputElement>) => void
}

export const Input = ({
	type = 'text',
	labelText,
	error,
	optional,
	placeholder,
	variant,
	ref,
	...props
}: Props) => {
	const containerClassName = clsx(
		error && styles.containerError,
		styles.container,
	)

	const inputClassName = clsx(
		{
			[styles.inputError]: Boolean(error),
		},
		styles[variant],
		styles.input,
	)

	return (
		<div>
			<div className={containerClassName}>
				<label className={styles.label} htmlFor={labelText}>
					{labelText} {optional && '(optional)'}
				</label>
				<input
					ref={ref}
					type={type}
					id={labelText}
					data-testid="input"
					placeholder={placeholder}
					className={inputClassName}
					{...props}
				/>
			</div>
			{error && <p className={styles.error}>{error}</p>}
		</div>
	)
}
