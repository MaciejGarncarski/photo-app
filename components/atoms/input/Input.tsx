import clsx from 'clsx';
import { ChangeEvent, forwardRef } from 'react';
import { FieldError } from 'react-hook-form';

import styles from './input.module.scss';

type PropsTypes = {
  type?: 'text' | 'number' | 'tel' | 'email' | 'password';
  labelText: string;
  error?: FieldError;
  optional?: boolean;
  className?: string;
  onChange?: (ev: ChangeEvent<HTMLInputElement>) => void;
  value?: string;
};

export const Input = forwardRef<HTMLInputElement, PropsTypes>(
  ({ type = 'text', labelText, error, optional, className, onChange, value, ...props }, ref) => {
    const containerClassName = clsx(className, error && styles.containerError, styles.container);

    const inputClassName = clsx(
      {
        [styles.inputError]: Boolean(error),
      },
      styles.input,
    );

    return (
      <div className={containerClassName}>
        <label className={styles.label} htmlFor={labelText}>
          {labelText} {optional && '(optional)'}
        </label>
        <input
          ref={ref}
          type={type}
          id={labelText}
          data-testid="input"
          className={inputClassName}
          onChange={onChange}
          value={value}
          {...props}
        />
        {error && <p className={styles.error}>{error.message}</p>}
      </div>
    );
  },
);
