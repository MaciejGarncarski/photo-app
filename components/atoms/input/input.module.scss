@use '/styles/common.scss';

.container {
  display: flex;
  flex-direction: column;
  position: relative;
  gap: 0.5rem;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  color: var(--text-secondary);
  text-transform: capitalize;
}


.input {
  background-color: transparent;
  border: none;
  padding: 1rem;
  border: var(--border-gray);
  border-radius: var(--radius-medium);
  transition: all var(--transition);
  transition-property: background-color, box-shadow;

  &:hover {
    background-color: var(--bg-secondary);
  }

  &:focus {
    outline: none;
    border-color: transparent;
    box-shadow: var(--shadow-focus);
  }
}

.inputError {
  background-color: var(--bg-error);
  border-color: var(--color-error);
}


.error {
  color: var(--color-error);
  font-weight: 600;
}

.textarea {
  border: none;
  padding: 1.5rem;
  max-width: calc(100% - 1px);
  resize: vertical;
}