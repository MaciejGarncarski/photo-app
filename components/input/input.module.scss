@use 'common';

.container {
  display: flex;
  flex-direction: column;
  position: relative;
  gap: 0.5rem;
  --gap: 0.7rem;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  color: var(--color-text-gray);
  transition: all var(--transition);
  transition-property: background-color, transform, opacity;
  display: inline-block;
}

.input {
  background-color: var(--color-secondary);
  padding: 1rem var(--gap);
  border-radius: var(--radius-small);
  border: 1px solid var(--shadow-0-5);
  box-shadow: var(--shadow);
  color: var(--color-text);

  &:focus {
    outline: none;
  }
}

.primary {
  background-color: var(--color-secondary);
}

.secondary {
  @include common.mediaquery('desktop') {
    background-color: var(--color-primary);
  }
}

.inputError {
  background-color: var(--color-error-light);

  &::placeholder {
    color: var(--color-text);
    font-weight: 500;
  }
}

.error {
  color: var(--color-error);
  padding-left: var(--gap);
  font-weight: 400;
  margin-left: 0.75rem;
  margin-top: 0.25rem;
}
