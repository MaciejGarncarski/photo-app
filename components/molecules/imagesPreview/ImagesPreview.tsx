import { IconPhotoPlus, IconTrash } from '@tabler/icons';
import { motion, Variants } from 'framer-motion';
import Image from 'next/image';

import { Heading } from '@/components/atoms/heading/Heading';
import { VisuallyHiddenText } from '@/components/atoms/visuallyHiddenText/VisuallyHiddenText';
import { ImagesBase64 } from '@/components/pages/createPost/types';

import styles from './imagesPreview.module.scss';

type PropsTypes = {
  onRemove: (id: string) => void;
  imagesBase64: ImagesBase64;
};

export const containerVariants: Variants = {
  hidden: { opacity: 0 },
  show: {
    opacity: 1,
    transition: {
      staggerChildren: 0.15,
    },
  },
};

export const itemVariants: Variants = {
  hidden: { opacity: 0, scale: 0.3, rotate: -4 },
  show: { opacity: 1, scale: 1, rotate: 0 },
};

export const ImagesPreview = ({ imagesBase64, onRemove }: PropsTypes) => {
  const emptyImagesLength = 3 - imagesBase64.length;

  return (
    <div>
      <Heading tag="h2">Images in post</Heading>
      <motion.div className={styles.previewContainer} variants={containerVariants} initial="hidden" animate="show">
        {imagesBase64.map((finalImage) => {
          if (finalImage?.src) {
            return (
              <motion.div key={finalImage.id} className={styles.previewButton} variants={itemVariants}>
                <motion.button
                  whileFocus={{ scale: 1.1 }}
                  whileHover={{ scale: 1.1 }}
                  whileTap={{ scale: 0.9 }}
                  type="button"
                  onClick={() => onRemove(finalImage.id)}
                  className={styles.deleteIcon}
                >
                  <IconTrash />
                  <VisuallyHiddenText text="remove image" />
                </motion.button>

                {imagesBase64 && (
                  <Image
                    className={styles.imgPreview}
                    src={finalImage.src}
                    alt="image preview"
                    width={200}
                    height={200}
                  />
                )}
              </motion.div>
            );
          }

          return null;
        })}

        {Array.from({ length: emptyImagesLength }, (_, el) => el).map((id) => {
          return (
            <motion.button variants={itemVariants} key={id} disabled type="button" className={styles.emptySpace}>
              <IconPhotoPlus />
              <p>empty</p>
            </motion.button>
          );
        })}
      </motion.div>
    </div>
  );
};
