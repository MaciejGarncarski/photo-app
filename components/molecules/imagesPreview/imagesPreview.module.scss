@use '/styles/common.scss';

.previewContainer {
  display: grid;
  place-content: center;
  grid-template-columns: repeat(3, 1fr);
  gap: 1rem;
  margin-top: 1rem;

  @include common.mediaquery('desktop') {
    gap: 3rem;
  }
}

.previewButton {
  border: none;
  background-color: transparent;
  display: flex;
  border-radius: var(--radius-small);
  box-shadow: var(--shadow-secondary);
  justify-content: center;
  align-items: center;
  border: var(--border-gray);
  position: relative;

  &:hover>.deleteIconOverlay {
    opacity: 1;
  }

}

.imgPreview {
  object-fit: contain;
  border-radius: var(--radius-small);
  width: 100%;
  height: auto;
}

.emptySpace {
  background-color: var(--color-primary);
  color: rgb(255, 255, 255);
  border-radius: var(--radius-small);
  border: 0;
  box-shadow: var(--shadow-secondary);
  display: flex;
  flex-direction: column;
  gap: 0.25rem;
  justify-content: center;
  align-items: center;
  min-height: 5rem;

  >svg {
    width: 2rem;
    height: 2rem;
  }

  @include common.mediaquery('desktop') {
    min-height: 8rem;
  }
}

.deleteIcon {
  position: absolute;
  right: -0.75rem;
  top: -1rem;
  background-color: var(--color-secondary);
  border-radius: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  color: var(--text-primary);
  padding: 0.3rem;
  border: none;

  &:hover,
  &:focus {
    outline-offset: 3px;
    background-color: var(--color-error);
  }

  >svg {
    color: rgb(20, 20, 20);
    width: 2rem;
    height: 2rem;
  }

  @include common.mediaquery('desktop') {
    right: -1.5rem;
    top: -1.5rem;

    >svg {
      width: 2.5rem;
      height: 2.5rem;
    }
  }
}