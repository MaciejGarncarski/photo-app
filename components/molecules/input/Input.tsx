import clsx from 'clsx';
import { forwardRef } from 'react';
import { FieldError } from 'react-hook-form';

import styles from './input.module.scss';

type InputProps = {
  type?: JSX.IntrinsicElements['input']['type'] | 'textarea';
  labelText: string;
  error?: FieldError;
  optional?: boolean;
  className?: string;
};

export const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ type = 'text', labelText, error, optional, className, ...props }, ref) => {
    const containerClassName = clsx(className, error && styles.containerError, styles.container);

    const inputClassName = clsx(
      {
        [styles.inputError]: Boolean(error),
      },
      styles.input,
    );

    return (
      <div className={containerClassName}>
        <label className={styles.label} htmlFor={labelText}>
          {labelText} {optional && '(optional)'}
        </label>
        <input ref={ref} type={type} id={labelText} data-testid="input" className={inputClassName} {...props} />
        {error && <p className={styles.error}>{error.message}</p>}
      </div>
    );
  },
);
