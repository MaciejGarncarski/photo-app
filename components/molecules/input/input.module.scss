@use '/styles/common.scss';

.container {
  display: flex;
  flex-direction: column;
  border: 2px solid var(--color-primary);
  border-radius: var(--radius-small);
  position: relative;
}

.containerError {
  border-color: var(--color-error);
}

.label {
  position: absolute;
  left: 0rem;
  top: 0;
  padding: 0 1rem;
  color: var(--text-placeholder);
  transition: all 100ms ease-out;
  cursor: text;
  transition-property: background-color, transform, font-size;
}

.labelActive,
.input:focus+.label {

  background-color: var(--bg-primary);
  color: var(--text-primary);
  top: -1.25rem;
  left: 0.5rem;


  @include common.mediaquery('lg') {
    font-size: 2.5rem;
  }

  @include common.mediaquery('xl') {
    top: -2rem;
    left: 1rem;
  }
}

.input {
  background-color: transparent;
  border: none;
  padding: 1rem;
}

.error {
  color: var(--color-error);
  font-weight: 600;
  font-size: 1.5rem;
  margin-left: 1rem;

  @include common.mediaquery('lg') {
    font-size: 2rem;
  }
}

.textarea {
  border: none;
  padding: 1.5rem;
  max-width: calc(100% - 1px);
  font-size: 2rem;
  resize: vertical;
}