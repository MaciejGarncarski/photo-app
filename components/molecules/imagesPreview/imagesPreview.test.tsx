import { fireEvent, render, screen } from '@testing-library/react';

import { ImagesPreview } from '@/components/molecules/imagesPreview/ImagesPreview';
import { snapshotComponent } from '@/components/molecules/input/input.test';

const img =
  'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII';

describe('Images preview', () => {
  describe('match snapshot', () => {
    snapshotComponent(<ImagesPreview imagesBase64={[undefined]} onRemove={jest.fn()} />);
  });
  describe('User integration', () => {
    it('Should detect onRemove', () => {
      const onRemove = jest.fn();
      render(
        <ImagesPreview
          imagesBase64={[
            {
              id: 0,
              src: img,
            },
          ]}
          onRemove={onRemove}
        />,
      );

      const imageDeleteBtn = screen.getByText(/remove image/);

      fireEvent.click(imageDeleteBtn);

      expect(onRemove).toHaveBeenCalled();
    });
  });
});
