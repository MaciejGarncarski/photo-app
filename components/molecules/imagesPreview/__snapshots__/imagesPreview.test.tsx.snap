// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <label
      class="label"
      for="random"
    >
      random 
    </label>
    <input
      class="input"
      data-testid="test"
      id="random"
      type="text"
    />
  </div>
</DocumentFragment>
`;

exports[` 2`] = `
<DocumentFragment>
  <div>
    <h2
      class="heading"
    >
      Images in post
    </h2>
    <div
      class="previewContainer"
      style="opacity: 0;"
    >
      <button
        class="emptySpace"
        disabled=""
        style="opacity: 0;"
        type="button"
      >
        <svg
          class="icon icon-tabler icon-tabler-photo-plus"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
            stroke="none"
          />
          <path
            d="M15 8h.01"
          />
          <path
            d="M12 20h-5a3 3 0 0 1 -3 -3v-10a3 3 0 0 1 3 -3h10a3 3 0 0 1 3 3v5"
          />
          <path
            d="M4 15l4 -4c.928 -.893 2.072 -.893 3 0l4 4"
          />
          <path
            d="M14 14l1 -1c.617 -.593 1.328 -.793 2.009 -.598"
          />
          <path
            d="M16 19h6"
          />
          <path
            d="M19 16v6"
          />
        </svg>
        <span
          class="visually-hidden"
        >
          empty space for image
        </span>
      </button>
      <button
        class="emptySpace"
        disabled=""
        style="opacity: 0;"
        type="button"
      >
        <svg
          class="icon icon-tabler icon-tabler-photo-plus"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
            stroke="none"
          />
          <path
            d="M15 8h.01"
          />
          <path
            d="M12 20h-5a3 3 0 0 1 -3 -3v-10a3 3 0 0 1 3 -3h10a3 3 0 0 1 3 3v5"
          />
          <path
            d="M4 15l4 -4c.928 -.893 2.072 -.893 3 0l4 4"
          />
          <path
            d="M14 14l1 -1c.617 -.593 1.328 -.793 2.009 -.598"
          />
          <path
            d="M16 19h6"
          />
          <path
            d="M19 16v6"
          />
        </svg>
        <span
          class="visually-hidden"
        >
          empty space for image
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
